deepinv/sampling/sde.py

```diff
@@ -0,0 +1,408 @@
+        :param str method: method for solving the SDE. One of the methods available in :meth:`deepinv.sampling.sde_solver`.
+
+        :return Tuple[Tensor, Tensor]: discretized drift and diffusion.
+        """
+        self.rng_manual_seed(seed)
+        solver_fn = select_solver(method)
+        solver = solver_fn(sde=self, rng=self.rng)
+        solution = solver.sample(x_init, timesteps=timesteps, *args, **kwargs)
+        return solution
+
+    def discretize(
+        self, x: Tensor, t: Union[Tensor, float], *args, **kwargs
+    ) -> Tuple[Tensor, Tensor]:
+        r"""
+        Discretize the SDE at the given time step.
+
+        :param torch.Tensor x: current state.
+        :param float t: discretized time step.
+        :param args: additional arguments for the drift.
+        :param kwargs: additional keyword arguments for the drift.
+
+        :return Tuple[Tensor, Tensor]: discretized drift and diffusion.
+        """
+        return self.drift(x, t, *args, **kwargs), self.diffusion(t)
+
+    def rng_manual_seed(self, seed: int = None):
+        r"""
+        Sets the seed for the random number generator.
+
+        :param int seed: the seed to set for the random number generator. If not provided, the current state of the random number generator is used.
+            Note: it will be ignored if the random number generator is not initialized.
+        """
+        if seed is not None:
+            if self.rng is not None:
+                self.rng = self.rng.manual_seed(seed)
+            else:
+                warnings.warn(
+                    "Cannot set seed for random number generator because it is not initialized. The `seed` parameter is ignored."
+                )
+
+    def reset_rng(self):
+        r"""
+        Reset the random number generator to its initial state.
+        """
+        self.rng.set_state(self.initial_random_state)
+
+    def randn_like(self, input: torch.Tensor, seed: int = None):
+        r"""
+        Equivalent to :func:`torch.randn_like` but supports a pseudorandom number generator argument.
+
+        :param torch.Tensor input: The input tensor whose size will be used.
+        :param int seed: The seed for the random number generator, if :attr:`rng` is provided.
+
+        :return: A tensor of the same size as input filled with random numbers from a normal distribution.
+        :rtype: torch.Tensor
+
+        This method uses the :attr:`rng` attribute of the class, which is a pseudo-random number generator
+        for reproducibility. If a seed is provided, it will be used to set the state of :attr:`rng` before
+        generating the random numbers.
+
+        .. note::
+           The :attr:`rng` attribute must be initialized for this method to work properly.
+        """
+        self.rng_manual_seed(seed)
+        return torch.empty_like(input).normal_(generator=self.rng)
+
+
+class DiffusionSDE(nn.Module):
```

Comment:
This shouldn't inherit from BaseSDE?

Reply:
This class has a forward and a backward SDE, meanwhile the BaseSDE class has only one process (either forward or backward). We can make this class inherit from the BaseSDE and do the same operations only for the backward SDE, but I think we can keep this independent of the BaseSDE class. To be discussed more in details.